druntime: Prepare `test/shared` for Windows support

kinke opened the pull request:
Upstreaming from LDC.

kinke committed:
druntime: Prepare test/shared source files for Windows support

Upstreaming from LDC.

kinke committed:
druntime: Prepare test/shared/Makefile for broader Windows support

maxhaton commented:
Should be one commit, no?

kinke commented on druntime/Makefile:
```diff
@@ -117,7 +117,6 @@ UTFLAGS:=-version=CoreUnittest -unittest -checkaction=context
 
 # Set PHOBOS_DFLAGS (for linking against Phobos)
 PHOBOS_PATH=../../phobos
-SHARED=$(if $(findstring $(OS),linux freebsd),1,)
```
This is already set in line 79, where it is (most likely correctly) enabled on DragonFly BSD too.

kinke commented on druntime/test/shared/Makefile:
```diff
@@ -37,41 +36,34 @@ $(ROOT)/dllgc$(DOTEXE): $(SRC)/dllgc.d
-TESTS+=link_linkdep load_linkdep link_loaddep load_loaddep load_13414
-
-EXPORT_DYNAMIC=$(if $(findstring $(OS),linux freebsd dragonflybsd),-L--export-dynamic,)
-NO_AS_NEEDED=$(if $(findstring $(OS),linux freebsd dragonflybsd),-L--no-as-needed,)
```
These 2 were unused.

kinke commented on druntime/test/shared/Makefile:
```diff
@@ -82,34 +74,42 @@ $(ROOT)/load$(DOTEXE) $(ROOT)/finalize$(DOTEXE): $(ROOT)/%$(DOTEXE): $(SRC)/%.d
 	$(QUIET)$(DMD) $(DFLAGS) -of$@ $< $(LINKDL)
 
-$(ROOT)/dynamiccast$(DOTEXE): $(SRC)/dynamiccast.d $(ROOT)/dynamiccast$(DOTDLL) $(DRUNTIMESO)
+$(ROOT)/dynamiccast$(DOTEXE): $(SRC)/dynamiccast.d $(ROOT)/dynamiccast$(DOTDLL) $(if $(LINK_SHARED),$(DRUNTIMESO),$(DRUNTIME))
```
The `dynamiccast` subtest is also enabled for Windows without `SHARED=1`, where it needs to depend on `DRUNTIME`, not `DRUNTIMESO`.

kinke commented:
> Should be one commit, no? At least the makefile adapting to the move should be in with the move from what I can see.

The 1st commit is independent (was green before being amended later, without CI checks since then due to 2nd commit) and an almost 1:1 upstreaming from LDC. The 2nd commit is a refactoring+generalization, which I'm testing at the same time with LDC and a few remaining mods in https://github.com/ldc-developers/ldc/pull/4563, for coverage on macOS and Windows.

kinke commented:
The conflict caused by #15974 is annoying; I wanted this to land first, exactly to prevent tests duplication on Windows (`link` here), as those barely need anything special for LDC (and so presumably ought to work with DMD's final DLL support on Windows too).

rainers commented on druntime/test/shared/Makefile:
```diff
@@ -82,34 +74,42 @@ $(ROOT)/load$(DOTEXE) $(ROOT)/finalize$(DOTEXE): $(ROOT)/%$(DOTEXE): $(SRC)/%.d
-	$(QUIET)$(DMD) -fPIC -shared $(DFLAGS) -of$@ $< $(LINKDL)
-
-endif # end Posix
+	$(QUIET)$(DMD) -shared $(DFLAGS) -of$@ $< $(LINKDL)
```
You can build link.exe and lib.dll for Windows, too, if you add `$(if $(findstring $(OS),windows),-visibility=public)` here (not implied yet by dmd with -shared). What is also needed is to replace `-dllimport=defaultLibsOnly` to `dllimport=all`.

rainers commented on druntime/test/shared/Makefile:
```diff
@@ -82,34 +74,42 @@ $(ROOT)/load$(DOTEXE) $(ROOT)/finalize$(DOTEXE): $(ROOT)/%$(DOTEXE): $(SRC)/%.d
 
 $(ROOT)/linkDR$(DOTEXE): $(SRC)/linkDR.c $(ROOT)/lib$(DOTDLL) $(DRUNTIMESO)
-	$(QUIET)$(CC) $(CFLAGS) -o $@ $< $(DRUNTIMESO) $(LDL) -pthread
+	$(QUIET)$(CC) $(CFLAGS) $(CC_OUTFLAG)$@ $(if $(findstring $(OS),windows),/Fo$@.obj,) $< $(DRUNTIMESO) $(CC_EXTRAS)
```
Use $(DRUNTIME_IMPLIB) instead of $DRUNTIMESO). Then linkD, linkDR and loadDR pass, too.

rainers reviewed (approved):
Looks good to me. If you rebase taking your versions in case of conflict, the mentioned changes make the `link` test pass again.

If you agree to the changes I can (force) push a rebased version.

kinke commented:
Thanks Rainer, much appreciated. For comparison, here's the remaining relevant diff of LDC (excl. an extra test…), for the `shared` Makefile, compared to this one here:
```diff
@@ -4,10 +4,30 @@
 include ../common.mak # affected by LINK_SHARED!
 
 ifneq (,$(LINK_SHARED))
-    # TODO: enable tests on Windows
-    ifneq (windows,$(OS))
-        TESTS:=link load linkD linkDR loadDR host finalize dynamiccast \
-               link_linkdep load_linkdep link_loaddep load_loaddep load_13414
+    # LDC: disable 3 tests on Mac, 1 on Windows
+    ifeq (,$(findstring ldmd2,$(DMD)))
+        # TODO: enable tests on Windows
+        ifneq (windows,$(OS))
+            TESTS:=link load linkD linkDR loadDR host finalize dynamiccast \
+                   link_linkdep load_linkdep link_loaddep load_loaddep load_13414
+        endif
+    else
+        TESTS:=link load linkD linkDR loadDR dynamiccast \
+               link_linkdep link_loaddep load_loaddep load_13414
+        ifneq ($(OS),osx)
+            # * `host` loads two modules with the same name, which is currently disallowed
+            #   by the (potentially overly eager) module collision detection on OS X.
+            # * `finalize` fails starting with macOS 10.13, as .dylibs with TLS can't be
+            #   unloaded anymore (https://github.com/ldc-developers/ldc/issues/3002).
+            # * FIXME: `load_linkdep`
+            #   it might fail because of unimplemented `getDependencies()` in rt.sections_elf_shared
+            ifeq (windows,$(OS))
+                # LDC FIXME: disable `load_linkdep` on Windows - needs `getDependencies()`
+                TESTS+=host finalize
+            else
+                TESTS+=host finalize load_linkdep
+            endif
+        endif
     endif
 endif
 # there are extra tests for Windows, not requiring a druntime DLL
@@ -22,8 +42,13 @@
 
 ifeq (windows,$(OS)) # extra tests on Windows
 
-ifeq ($(SHARED),1)
-DFLAGS+=-version=SharedRuntime
+ifneq (,$(findstring ldmd2,$(DMD)))
+    # LDC: required for executables to implicitly dllimport DLL data symbols
+    DFLAGS+=-dllimport=all
+else
+    ifeq ($(SHARED),1)
+        DFLAGS+=-version=SharedRuntime
+    endif
 endif
 
 $(ROOT)/dllrefcount$(DOTEXE): $(SRC)/dllrefcount.d
@@ -32,6 +57,11 @@
 $(ROOT)/loadlibwin$(DOTEXE): $(SRC)/loadlibwin.d
 	$(QUIET)$(DMD) $(DFLAGS) -of$@ $<
 
+# LDC: this test is designed for .exe & .dll with separate druntimes
+ifneq (,$(findstring ldmd2,$(DMD)))
+$(ROOT)/dllgc$(DOTEXE): DFLAGS+=-link-defaultlib-shared=false -dllimport=none
+endif
+
 $(ROOT)/dllgc$(DOTEXE): $(SRC)/dllgc.d
 	$(QUIET)$(DMD) $(DFLAGS) -version=DLL -shared -of$(ROOT)/dllgc$(DOTDLL) $<
 	$(QUIET)$(DMD) $(DFLAGS) -of$@ $<
```

So I'm adding `-dllimport=all` here on Windows, for all tests except the `dllgc` one (which also links static druntime). IIRC, the test worked fine, but explicitly tests the GC proxy with multiple binaries containing a static druntime each. I don't think I've checked whether `-dllimport=all` is really required for all other tests; but if DMD needs it too, then great, no need to special-case anymore.

The weird current behavior with DMD on Windows is that `test/shared` is tested 'regularly' with `SHARED=` (i.e., static druntime), and another time manually with `SHARED=1` for one set of Azure CI jobs. So this currently requires the `SharedRuntime` version, and probably makes enabling the remaining tests a bit trickier.

And wrt. `DRUNTIME_IMPLIB` - I guess that's fine, but I'm simply using the `.lib` as `DRUNTIMESO` on Windows. But I also don't have to link it explicitly in DFLAGS, I simply use `-link-defaultlib-shared` (in `common.mak`, for `LINK_SHARED=1`).

rainers commented:
I guess some of the issues are caused by the DLL support in DMD not being complete yet: `-version=SharedRuntime` is only needed because exceptions don't work reliably, something like -link-defaultlib-shared is probably needed, too, but requires a built and working phobos_shared.dll.

rainers committed:
test/shared for windows:
- add -visibility=public when building DLLs
- use -dllimport=all to simplify building exe
- enable tests link linkD linkDR loadDR

rainers commented:
Sorry @kinke , got distracted. Now rebased again and force-pushed minor changes. The DFLAGS for DLLs now get an additional -visibility=public that should just set the same as -shared for LDC.

kinke commented on druntime/test/shared/Makefile:
```diff
@@ -1,34 +1,38 @@
 
-TESTS:=loadlibwin dllrefcount dllgc dynamiccast
 ifeq ($(SHARED),1)
 TESTS+=link
```
I think this is superfluous now, already added in line 9 (`LINK_SHARED=SHARED`). And adding to TESTS potentially too late here anyway. [line re-added after rebase]

kinke replied:
Also re-added: prepending to PATH in line 33. I was pretty sure this wouldn't have any effect without `export`.

kinke replied:
Then there's a new Windows-specific `SimpleDllMain` mixin in `druntime/test/shared/src/lib.d` in master; ideally that'd be DMD-specific, to make clear that the other compilers don't need such extra stuff on Windows.

Thanks Rainer!

kinke replied:
> Also re-added: prepending to PATH in line 33. I was pretty sure this wouldn't have any effect without export.

Oh well, looks like that somehow works without `export` and is needed.

kinke committed:
[nits after rebase]

rainers committed:
pass -visibility=public for SHARED=1

WalterBright commented:
This is producing a blizzard of warnings from loadDR.c in Windows_DMD_bootstrap x64 tests.

```
src/loadDR.c(13): warning C4152: nonstandard extension, function/data pointer conversion in expression
src/loadDR.c(14): warning C4152: nonstandard extension, function/data pointer conversion in expression
src/loadDR.c(15): warning C4152: nonstandard extension, function/data pointer conversion in expression
src/loadDR.c(16): warning C4152: nonstandard extension, function/data pointer conversion in expression
src/loadDR.c(30): warning C4152: nonstandard extension, function/data pointer conversion in expression
D:\a\1\s\druntime\test\shared\src\loadDR.c(10) : warning C4711: function 'loadLib' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(13) : warning C4711: function 'loadSym' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(14) : warning C4711: function 'loadSym' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(15) : warning C4711: function 'loadSym' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(16) : warning C4711: function 'loadSym' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(30) : warning C4711: function 'loadSym' selected for automatic inline expansion
D:\a\1\s\druntime\test\shared\src\loadDR.c(37) : warning C4711: function 'closeLib' selected for automatic inline expansion
```
@kinke

WalterBright commented:
https://github.com/dlang/dmd/pull/16609#issuecomment-2183836567
